Doc/reference/expressions.rst

```diff
@@ -823,12 +823,18 @@ An attribute reference is a primary followed by a period and a name:
 
 The primary must evaluate to an object of a type that supports attribute
 references, which most objects do.  This object is then asked to produce the
-attribute whose name is the identifier.  This production can be customized by
-overriding the :meth:`__getattr__` method.  If this attribute is not available,
-the exception :exc:`AttributeError` is raised.  Otherwise, the type and value of
-the object produced is determined by the object.  Multiple evaluations of the
-same attribute reference may yield different objects.
-
+attribute whose name is the identifier. The type and value produced is
+determined by the object.  Multiple evaluations of the same attribute
+reference may yield different objects.
+
+This production can be customized by overriding the
+:meth:`~object.__getattribute__` method or the :meth:`~object.__getattr__`
+method.  The :meth:`!__getattribute__` method is called first and either
+returns a value or raises :exc:`AttributeError` if the attribute is not
+available.
```

Comment:
Of course, it _is_ possible to create badly behaved objects that raise other exceptions from `__getattribute__`:

```pycon
>>> class Foo:
...     def __getattribute__(self, attr):
...         if attr == 'foo':
...             raise TypeError('no')
...         return object.__getattribute__(self, attr)
...
>>> f = Foo()
>>> f.a = 1
>>> f.a
1
>>> f.b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    f.b
  File "<stdin>", line 5, in __getattribute__
    return object.__getattribute__(self, attr)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Foo' object has no attribute 'b'
>>> f.foo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    f.foo
  File "<stdin>", line 4, in __getattribute__
    raise TypeError('no')
TypeError: no
```

Does that need to be mentioned here? Or is it too obscure?

Reply:
No need.  Any code can raise an weird error at any time.   This wording corresponds to what the code actually does